File: password/solutions/password.js
process.stdin.resume();
process.stdin.setEncoding("utf-8");

let inputString = "";
let currentLine = 0;

process.stdin.on("data", (inputStdin) => {
    inputString += inputStdin;
});

process.stdin.on("end", () => {
    inputString = inputString.split("\n");
    main();
});

function readLine() {
    return inputString[currentLine++];
}
/**
 * @param {string[]} strings - An array of encoded strings.
 * @returns {string} - The deciphered password.
 */
function decipher(strings) {
    return strings
        .map((string) => {
            const mid = string.length / 2;
            const firstHalf = new Set(string.slice(0, mid));
            const secondHalf = new Set(string.slice(mid));

            const intersection = [...firstHalf].filter((char) =>
                secondHalf.has(char),
            );

            return intersection[0];
        })
        .join("");
}

function main() {
    const strings = [];
    const n = parseInt(readLine());
    for (let i = 0; i < n; i++) strings.push(readLine());
    console.log(decipher(strings));
}

File: password/solutions/password.cpp
#include <bits/stdc++.h>
using namespace std;

string decipher(vector<string> &strings) {
  string result = "";

  for (string &str : strings) {
    int mid = str.length() / 2;

    set<char> first_half(str.begin(), str.begin() + mid);
    set<char> second_half(str.begin() + mid, str.end());

    set<char> intersection;
    set_intersection(first_half.begin(), first_half.end(), second_half.begin(),
                     second_half.end(),
                     inserter(intersection, intersection.begin()));
    if (intersection.empty()) {
      continue;
    } else {
      result += *intersection.begin();
    }
  }

  return result;
}

int main() {
  int n;
  cin >> n;

  vector<string> strings(n);
  for (int i = 0; i < n; i++)
    cin >> strings[i];

  cout << decipher(strings) << endl;
  return 0;
}
